skrub/_agg_joiner.py

```diff
@@ -0,0 +1,597 @@
+        if not isinstance(self.table, (list, tuple)):
+            tables = [self.table]
+        else:
+            tables = self.table
+
+        if len(self.suffix_) == 0:
+            if len(tables) == 1:
+                self.suffix_ = [""]
+            else:
+                self.suffix_ = [f"_{idx+1}" for idx in range(len(tables))]
+
+        # Check tables and list of suffix match
+        if len(tables) != len(self.suffix_):
+            raise ValueError(
+                "'suffix' must be None or match the "
+                f"number of tables, got: {self.suffix_!r}"
+            )
+
+        # Check tables and list of foreign_keys match
+        if len(tables) != len(self.foreign_key_):
+            error_msg = (
+                "The number of tables must match the number of foreign_key, "
+                f"got {len(tables)=!r} and {len(self.foreign_key_)=!r}. "
+            )
+            if len(tables) > 1:
+                error_msg += (
+                    "For multiple tables, use a list of list, "
+                    "e.g. foreign_key=[['col1', 'col2'], ['colA', 'colB']]."
+                )
+            raise ValueError(error_msg)
+
+        # Check table type and missing columns
+        for idx, (table, foreign_key, cols) in enumerate(
+            zip(tables, self.foreign_key_, self.cols_)
+        ):
+            if isinstance(table, str):
+                if table != "X":
+                    raise ValueError(
+                        "If the dataframe is declared with a string, "
+                        f"the only acceptable value is 'X', got {table!r}."
+                    )
+                table = deepcopy(X)
+                tables[idx] = table
+
+            elif not hasattr(table, "__dataframe__"):
+                raise TypeError(
+                    "'tables' must be a list of tuple and the first element of each"
+                    f" tuple must be a dataFrame, got {type(tables[0])} at index"
+                    f" {idx}."
+                )
+
+            # If no cols provided, all columns but foreign_key are used.
+            if len(cols) == 0:
+                cols = list(set(table.columns) - set(foreign_key))
+                self.cols_[idx] = cols
+
+            error_msg = f"{foreign_key=!r} are not in {table.columns=!r}."
+            check_missing_columns(table, foreign_key, error_msg=error_msg)
+
+            error_msg = f"{cols=!r} are not in {table.columns=!r}."
+            check_missing_columns(table, cols, error_msg=error_msg)
+
+            if len(foreign_key) != len(self.main_key_):
+                raise ValueError(
+                    "The number of keys to join must match, got "
+                    f"main_key={self.main_key_!r} and "
+                    f"{foreign_key=!r} for the table at index {idx}."
+                )
+
+        self.table_ = tables
+
+        # Check tables and list of cols match
+        if len(tables) != len(self.cols_):
```

Comment:
the amount of validation needed and the complexity of the parameter makes me wonder if we should allow only a single auxiliary table? In general I find parameters that can be str, list[str] or list[list[str]] a bit confusing because here if I want a single column per table I might expect to be able to pass them as a list of str, since I am allowed to pass a str for one table

Reply:
Yes that's a good point. Our perspective is that running agg joiner or joiner on multiple aux tables is probably quite niche, but still something we'd like to have, when working with many tables from a datalake for instance.

So using a list of list allows to makes things less ambiguous as a whole, but I agree that the params validation starts to get complex.

Reply:
that makes sense. in this case could we have `main_key` and `foreign_key` have a plural name and force them to be a list of str for each table? this way we would have only 2 options: `list[list[str]]` for multiple tables or `list[str]` for a single table, and it would remove any ambiguity of the interpretation of `list[str]`

Reply:
also comparing lengths might fail if someone has 2 tables and 2 grouping columns

Reply:
> that makes sense. in this case could we have `main_key` and `foreign_key` have a plural name and force them to be a list of str for each table? this way we would have only 2 options: `list[list[str]]` for multiple tables or `list[str]` for a single table, and it would remove any ambiguity of the interpretation of `list[str]`

I would agree with @gvaroquaux here that we want the user to be able to use simple strings. The ambiguity is limited to the case of multiple tables, raising errors when the format is incorrect.

If this is an issue, I'd rather drop the support of multiple tables, and maybe create a specific estimator like MultiAggJoiner.

WDYT?

Reply:
> also comparing lengths might fail if someone has 2 tables and 2 grouping columns

I think we already handle this situation

Reply:
ah yes you're right, with the `atleast_2d_or_none` call

Reply:
> > that makes sense. in this case could we have `main_key` and `foreign_key` have a plural name and force them to be a list of str for each table? this way we would have only 2 options: `list[list[str]]` for multiple tables or `list[str]` for a single table, and it would remove any ambiguity of the interpretation of `list[str]`
> 
> I would agree with @gvaroquaux here that we want the user to be able to use simple strings. The ambiguity is limited to the case of multiple tables, raising errors when the format is incorrect.
> 
> If this is an issue, I'd rather drop the support of multiple tables, and maybe create a specific estimator like MultiAggJoiner.
> 
> WDYT?

I like the MultiAggJoiner idea! I find it hard when a parameter can take many similar forms with different meanings so splitting the functionality like this might be a good thing. but if you both prefer the current interface it makes sense too!

Reply:
An alternative could be to accept only one table but provide a method to add more:

```python
# constructor used by all users

joiner = AggJoiner(table=aux, foreign_key="from_airport", main_key="airportId")

# method to add more columns, used by more advanced users who need to join many tables

joiner.add_table(table=aux1, foreign_key="to_airport", main_key="airportId")

# or if we really want the list of tables

joiner.add_tables(tables=[...], ...)

# to avoid having to awkwardly pass the first table to the constructor and then
# add the rest in the advanced use case we could allow joiner=AggJoiner(),
# creating an empty joiner that doesn't do anything unless we call add_table
# later
```

or have a classmethod like `from_tables`